[12.x] feat: Add Arr::date helper to retreive Carbon instances

DarronEngelbrecht opened the pull request:
This PR introduces an `Illuminate\Support\Arr::date()` method.

It attempts to match `request()->date('input')` functionality for arrays, using  `Arr::get()` to fetch the key or array and retrieving results as Date instances, or null for invalid values or missing keys. 

Example usage
```php
$array = ['due_date' => '2023-03-12'];

$date = Arr::date(
    array: $array, 
    key: 'due_date', 
    format: 'Y-m-d',
    tz: 'GMT'
 );

// previously I would do something like this.. 
$array = ['due_date' => '2023-03-12'];
try { 
   $dueDate = Carbon::parse(Arr::get($array, 'due_date')));
} catch (InvalidFormatException) { 
   $dueDate = null;
}

// now...
$dueDate = Arr::date($array, 'due_date');

//previously
$array = ['2025-02-12', '2025-03-15'];
$dates = array_map(function($date) {
try { 
   return Carbon::parse(Arr::get($date, 'due_date')));
} catch (InvalidFormatException) { 
   return null;
}}, $arr)

// now...
$array = ['2025-02-12', '2025-03-15'];
$dates = Arr::date($array, null);
```

Looking for opinions on existing easier ways to do this and returning null values vs throwing exceptions?

I guess it would be nice to have something like `Arr::tryDate` which should return null values where this would then throw exceptions?

Thoughts or concerns? 

Have PR'ed something for Carbon too: https://github.com/briannesbitt/Carbon/pull/3179 - can clean up here if that gets added, or add this to the Illuminate/Support/Carbon instance? 



<!--
Please only send a pull request to branches that are currently supported: https://laravel.com/docs/releases#support-policy 

If you are unsure which branch your pull request should be sent to, please read: https://laravel.com/docs/contributions#which-branch

Pull requests without a descriptive title, thorough description, or tests will be closed.

In addition, please describe the benefit to end users; the reasons it does not break any existing features; how it makes building web applications easier, etc.
-->

DarronEngelbrecht committed:
Arr date

DarronEngelbrecht committed:
Linting

DarronEngelbrecht committed:
Linting

DarronEngelbrecht committed:
Linting

rodrigopedra commented:
I would use the `Date` façade instead of the upstream `Carbon\Carbon`. So if one configures their project to use an immutable date factory, they wouldn't be surprised.

Or at least the `Illuminate\Support\Carbon` subclass.

DarronEngelbrecht committed:
Date facade and timezones

DarronEngelbrecht committed:
Added format

DarronEngelbrecht commented:
Thanks @rodrigopedra for the input.

Have had a look at `request()->date('input')` for answers.

1. Now returning the Date facade.
2. Added timezone input.
3. Added the format input.
4. Removed the default input. 

This more closely matches `request()->date()` and should work for a wider audience.

taylorotwell commented:
Thanks for your pull request to Laravel!

Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include. 

If applicable, please consider releasing your code as a package so that the community can still take advantage of your contributions!
